src/installer/__main__.py

```diff
@@ -0,0 +1,86 @@
+"""Installer CLI."""
+
+import argparse
+import distutils.dist
+import sys
+import sysconfig
+from typing import Dict, Optional, Sequence
+
+import installer
+import installer.destinations
+import installer.sources
+import installer.utils
+
+
+def main_parser() -> argparse.ArgumentParser:
+    """Construct the main parser."""
+    parser = argparse.ArgumentParser()
+    parser.add_argument("wheel", type=str, help="wheel file to install")
+    parser.add_argument(
+        "--destdir",
+        "-d",
+        metavar="/",
+        type=str,
+        help="destination directory (prefix to prepend to each file)",
+    )
+    parser.add_argument(
+        "--compile-bytecode",
+        action="append",
+        metavar="level",
+        type=int,
+        choices=[0, 1, 2],
+        help="generate bytecode for the specified optimization level(s) (default=0, 1)",
+    )
+    parser.add_argument(
+        "--no-compile-bytecode",
+        action="store_true",
+        help="don't generate bytecode for installed modules",
+    )
+    return parser
+
+
+def get_scheme_dict(distribution_name: str) -> Dict[str, str]:
+    """Calculate the scheme disctionary for the current Python environment."""
+    scheme_dict = sysconfig.get_paths()
+
+    # calculate 'headers' path, sysconfig does not have an equivalent
+    # see https://bugs.python.org/issue44445
+    dist_dict = {
+        "name": distribution_name,
+    }
+    distribution = distutils.dist.Distribution(dist_dict)
+    install_cmd = distribution.get_command_obj("install")
```

Comment:
FYI, this doesn't seem future-proof — between distutils being deprecated and setuptools deprecating install command.

Reply:
It's certainly not, but I don't think there's any alternative way to get the `headers` destination directory described in the wheel spec. There's a discussion about this here: https://discuss.python.org/t/clarification-on-a-wheels-header-data/9305 and a Python bug open: https://bugs.python.org/issue44445

I've so far not found any wheel that actually uses the headers directory - if that's generally true, then maybe the resolution for the future will be just to remove the `header` directory from the spec, so we can get all the paths from `sysconfig`. Otherwise, it will have to be added to the paths `sysconfig` knows about.

Reply:
`greenlet` is one that I use that has a headers folder.

Reply:
Aha, so it is used...

Reply:
FWIW, pip uses distutils for system installs under Python < 3.10, sysconfig in 3.10 and above, and it synthesises an include path for virtual envs under `sys.prefix`.

Reply:
Ugh. So if I'm reading pip's code right, the decision of whether to use paths from sysconfig or distutils is [this code](https://github.com/pypa/pip/blob/7f8a6844037fb7255cfd0d34ff8e8cf44f2598d4/src/pip/_internal/locations/__init__.py#L228-L246), and the calculation of a header scheme directory from sysconfig paths is [this code](https://github.com/pypa/pip/blob/7f8a6844037fb7255cfd0d34ff8e8cf44f2598d4/src/pip/_internal/locations/_sysconfig.py#L173-L191).

I don't think `installer`, as a low-level tool, should be detecting 'am I running in a virtualenv' - encoding high-level concepts in a low-level package like that feels wrong. It also means things behave differently in environments made with e.g. conda or pyenv, which I think goes against user expectations.

On my system, for a standard Posix prefix install of a distribution called `foobar`:

- distutils gives `{prefix}/include/python3.10/foobar` (both with and without a venv)
- pip's sysconfig logic outwith a virtualenv would give `{prefix}/include/python3.10/foobar`
- pip's sysconfig logic inside a virtualenv would give `{prefix}/include/site/python3.10/foobar`

The code in pip explicitly notes that it's somewhat arbitrary and only done for backwards compatibility. For a specific detail like this, I don't think installer should be bound to duplicate pip's behaviour exactly, so I'd suggest we ignore virtualenvs and always use `os.path.join(include_path, distribution_name)` as the header target. I'd guess that's also what an old-school `setup.py install` would do, though I haven't checked that.

Reply:
I've implemented the what distutils appears to do (make a subdirectory with the distribution name under `include`), but normalising the name like we've agreed to do for other places a distribution name is used to make a file or folder name (wheel filenames, `.dist-info` folders).

@pradyunsg I suspect there's more work to be done on where `installer` uses original vs normalised distribution names. At the moment, the distribution name is parsed out of the wheel filename, but this can either be partially or fully normalised. If we want the original name for anything, that needs to be read from the metadata inside the wheel.

Reply:
Let's leave this normalisation code out of the CLI, and file a follow up issue for API to address it in a blanket move.

Reply:
OK, done.

Reply:
Opened #97 for the name question.